src/geotop/snow.cc

```diff
@@ -1594,29 +1594,31 @@ double interpolate_snow(long r, long c, double h, long max, Tensor<double> *Dz,
 void copy_snowvar3D(STATEVAR_3D *from, STATEVAR_3D *to)
 {
 
-  long l, r, c;
-  long nl=from->Dzl->ndh, nr=from->Dzl->nrh, nc=from->Dzl->nch;
-
-  for (r=1; r<=nr; r++)
-  {
-    for (c=1; c<=nc; c++)
-    {
-      (*to->type)(r,c) = (*from->type)(r,c);
-      (*to->lnum)(r,c) = (*from->lnum)(r,c);
-      for (l=1; l<=nl; l++)
-      {
-        (*to->Dzl)(l,r,c) = (*from->Dzl)(l,r,c);
-        (*to->w_liq)(l,r,c) = (*from->w_liq)(l,r,c);
-        (*to->w_ice)(l,r,c) = (*from->w_ice)(l,r,c);
-        (*to->T)(l,r,c) = (*from->T)(l,r,c);
-      }
+    long l, r, c;
+    long nl=from->Dzl->ndh, nr=from->Dzl->nrh, nc=from->Dzl->nch;
+
+#pragma omp parallel for firstprivate(from, nr, nc), private(r,c)
```

Comment:
`from, nr, nc` should be **shared**. this is the recommended way (see e.g., https://mitpress.mit.edu/books/using-openmp)

Reply:
I've put them **firsprivate** because they have a value before entring in the parallel region, and the threads will modify some of the variables belonging to  ```from``` (i.e. ```type, lnum```). Maybe it works also using shared because those variables themselves are not modifed by threads.

Reply:
note that the value of a variable declared `firstprivate` is **undefined** at the end of the parallel region.. so if their valued is used also outside the parallel region, mark them as `shared`

Reply:
Oh, I didn't read this. From the output of my toycode it didn't seem like you said.
```
#include <iostream>
#include <omp.h>

int main(){
  
  int i = 10;
  
#pragma omp parallel firstprivate(i)
  {
    std::cout << "thread " << omp_get_thread_num() << " => i = " << i << std::endl;
    i = 1000 + omp_get_thread_num();
  }
  
  std::cout << "i = " << i << std::endl;
    
  std::cout << "i+5 = " << i+5 << std::endl;
  
  return 0;
}
```
Output:
```
thread 2 => i = 10
thread 7 => i = 10
thread 5 => i = 10
thread 3 => i = 10
thread 1 => i = 10
thread 6 => i = 10
thread 0 => i = 10
thread 4 => i = 10
i = 10
i+5 = 15
```

Reply:
the standard says it is undefined. meaning that each implementer can do as they prefer. please make them shared

Reply:
May I ask you a free source? 
I've downloaded the **OpenMP Application Programming Interface, Version 5.0 November 2018** and I couldn't find it, maybe I didn't search in the right place.

If you're sure about it, I'll make them shared.

Reply:
I am sure.

it is unlikely your compiler supports omp-5.0. 

here is the link of omp-4.5 
https://www.openmp.org/wp-content/uploads/openmp-4.5.pdf

page 193: it says that the content of a **private** variable is undefined after the region.

page 196: says that **firstprivate** enrich the private by providing a defined initial vale. All the rest is like private (meaning what I said)

if you want to retain a value of a private variable after the parallel region, use **lastprivate**. If it were like you said, **lastprivate** shouldn't be invented